src/System.Memory/src/System/Buffers/Text/Base64Encoder.cs

```diff
@@ -0,0 +1,231 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace System.Buffers.Text
+{
+    /// <summary>
+    /// Convert between binary data and UTF-8 encoded text that is represented in base 64.
+    /// </summary>
+    public static partial class Base64
+    {
+        /// <summary>
+        /// Encode the span of binary data into UTF-8 encoded text represented as base 64.
+        ///
+        /// <param name="bytes">The input span which contains binary data that needs to be encoded.</param>
+        /// <param name="utf8">The output span which contains the result of the operation, i.e. the UTF-8 encoded text in base 64.</param>
+        /// <param name="consumed">The number of input bytes consumed during the operation. This can be used to slice the input for subsequent calls, if necessary.</param>
+        /// <param name="written">The number of bytes written into the output span. This can be used to slice the output for subsequent calls, if necessary.</param>
+        /// <param name="isFinalBlock">True (default) when the input span contains the entire data to decode. 
+        /// Set to false only if it is known that the input span contains partial data with more data to follow.</param>
+        /// <returns>It returns the OperationStatus enum values:
+        /// - Done - on successful processing of the entire input span
+        /// - DestinationTooSmall - if there is not enough space in the output span to fit the encoded input
+        /// - NeedMoreData - only if isFinalBlock is false, otherwise the output is padded if the input is not a multiple of 3
+        /// It does not return InvalidData since that is not possible for base 64 encoding.</returns>
+        /// </summary> 
+        public static OperationStatus EncodeToUtf8(ReadOnlySpan<byte> bytes, Span<byte> utf8, out int consumed, out int written, bool isFinalBlock = true)
+        {
+            ref byte srcBytes = ref bytes.DangerousGetPinnableReference();
+            ref byte destBytes = ref utf8.DangerousGetPinnableReference();
+
+            int srcLength = bytes.Length;
+            int destLength = utf8.Length;
+
+            int sourceIndex = 0;
+            int destIndex = 0;
+            int result = 0;
+
+            ref byte encodingMap = ref s_encodingMap[0];
+
+            if (srcLength <= MaximumEncodeLength && destLength >= GetMaxEncodedToUtf8Length(srcLength))
```

Comment:
Would it make sense to cap the srcLength to whatever fits instead of having two copies of the encoding loop? It would avoid the penalty for the case where the destination buffer does not fit, and make the code quite a bit smaller (note that you get pretty non-trivial code expansion here because of the use of the AggressiveInlining attributes).

Reply:
Same in the decoder.

Reply:
Great idea. I was tinkering with that but was previously struggling to get the "cap" set correctly. But I believe I have figured it out now!